### ISSUE TITLE
[Bug]: "InvalidOperationException: Collection was modified" When using joins

### ISSUE THREAD

**Issue by rcm-adrien-gervaise** [reactivemarbles/DynamicData#787]
### Describe the bug 🐞

This is the come-back of #364  
Subscribing to a join is still not thread-safe, as GetInitialUpdates will enumerate the underlying Dictionary, which could be updated at the same time. 

### Step to reproduce

```
[Fact]
    public void LetsSeeWhetherWeCanRandomlyHitADifferentRaceCondition()
    {
        var leftCache = new SourceCache<Thing, long>(x => x.Id);
        var rightCache = new SourceCache<Thing, long>(x => x.Id);
        var joined = leftCache.Connect().InnerJoin(rightCache.Connect(), x => x.Id,
            (keys, leftThing, rightThing) =>
                new Thing { Id = keys.leftKey, Name = $"{leftThing.Name} x {rightThing.Name}" });

        IDisposable StartUpdating(ISourceCache<Thing, long> sourceCache) =>
            Observable.Range(1, 1000000, Scheduler.Default)
                .Subscribe(x => sourceCache.AddOrUpdate(new Thing { Id = x, Name = $"{x}" }));

        using var leftUpdater = StartUpdating(leftCache);
        using var rightUpdater = StartUpdating(rightCache);
        using var subscription = joined.Subscribe();
    }
```

### Reproduction repository

https://github.com/rcm-adrien-gervaise/DynamicData

### Expected behavior

No exception should be thrown

### Additional information ℹ️

Test reproducing the issue and suggested fix on https://github.com/rcm-adrien-gervaise/DynamicData

**Comment by rcm-adrien-gervaise** [comment]
No, this has been around for a while. We have the issue using 7.4.9, but I think it predates that.

**Comment by glennawatson** [comment]
> No, this has been around for a while. We have the issue using 7.4.9, but I think it predates that.

Just confirming it happens in the brand new release from yesterday?

**Comment by RolandPheasant** [comment]
That's a good reproduction thanks:

![image](https://github.com/reactivemarbles/DynamicData/assets/9879866/654e3088-ce59-4f05-9d0e-40c95163a1a4)

**Comment by RolandPheasant** [comment]
This has been fixed and deployed in v 8.3.27

**Comment by rcm-adrien-gervaise** [comment]
Thanks for the quick fix !
